File: Editorial/1756_Design_Most_Recently_Used_Queue/Java/3_Square_Root_Decomposition.java
class MRUQueue {

    private int totalElements;
    private int BUCKET_SIZE;
    List<List<Integer>> data = new ArrayList<>();
    List<Integer> index = new ArrayList<>();

    public MRUQueue(int n) {
        this.totalElements = n;
        this.BUCKET_SIZE = (int) Math.sqrt(n);
        for (int number = 1; number <= n; ++number) {
            int bucketIndex = (number - 1) / BUCKET_SIZE;
            if (bucketIndex == data.size()) {
                data.add(new ArrayList<>());
                index.add(number);
            }
            data.get(data.size() - 1).add(number);
        }
    }

    public int fetch(int k) {
        int bucketIndex = upperBound(index, k) - 1;
        int element = data.get(bucketIndex).get(k - index.get(bucketIndex));
        data.get(bucketIndex).remove(k - index.get(bucketIndex));

        for (int i = bucketIndex + 1; i < index.size(); ++i) {
            index.set(i, index.get(i) - 1);
        }

        if (data.get(data.size() - 1).size() >= BUCKET_SIZE) {
            data.add(new ArrayList<>());
            index.add(totalElements);
        }
        data.get(data.size() - 1).add(element);

        if (data.get(bucketIndex).isEmpty()) {
            data.remove(bucketIndex);
            index.remove(bucketIndex);
        }
        return element;
    }

    public int upperBound(List<Integer> nums, int target) {
        int left = 0, right = nums.size();
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums.get(mid) > target) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
}

File: Editorial/1701-1800/1756_Design_Most_Recently_Used_Queue/CPP/3_Square_Root_Decomposition.cpp
class MRUQueue {
private:
    int totalElements;
    const int BUCKET_SIZE;
    // Stores elements in different buckets
    vector<vector<int>> data;
    // Tracks the starting index of each bucket
    vector<int> index;

public:
    MRUQueue(int n) : totalElements(n), BUCKET_SIZE(sqrt(n)) {
        // Initialize data with the given number of elements
        for (int number = 1; number <= n; ++number) {
            int bucketIndex = (number - 1) / BUCKET_SIZE;
            // Create a new bucket if it's not already present
            if (bucketIndex == data.size()) {
                data.push_back({});
                index.push_back(number);
            }
            // Add the number to the corresponding bucket
            data.back().push_back(number);
        }
    }

    int fetch(int k) {
        // Find the bucket that contains the kth element using binary search
        int bucketIndex =
            upper_bound(index.begin(), index.end(), k) - index.begin() - 1;

        // Get the element from the selected bucket
        int element = data[bucketIndex][k - index[bucketIndex]];
        // Remove the element from its current bucket
        data[bucketIndex].erase(data[bucketIndex].begin() +
                                (k - index[bucketIndex]));

        // Shift the starting indices of the subsequent buckets
        for (int i = bucketIndex + 1; i < index.size(); ++i) {
            index[i]--;
        }

        // Append the element to the last bucket
        if (data.back().size() >= BUCKET_SIZE) {
            data.push_back({});
            // Update the last bucket's index
            index.push_back(totalElements);
        }
        // Add the element to the last bucket
        data.back().push_back(element);

        // Remove any empty buckets after the operation
        if (data[bucketIndex].empty()) {
            data.erase(data.begin() + bucketIndex);
            index.erase(index.begin() + bucketIndex);
        }

        return element;
    }
};
